src/app/server/Server.cpp

```diff
@@ -298,6 +298,20 @@ void Server::DispatchShutDownAndStopEventLoop()
     chip::DeviceLayer::PlatformMgr().ScheduleWork(StopEventLoop);
 }
 
+void Server::FactoryReset()
+{
+    // Delete all fabrics and emit Leave event.
```

Comment:
These functions should be called from the CHIP thread while you seem to assume that `FactoryReset` can be called from an arbitrary thread. I suggest:
1. Wrapping the functionality with `ScheduleWork`.
2. Renaming `FactoryReset` to `ScheduleFactoryReset` to make it more clear that the function doesn't do the job immediately.

Reply:
I was afraid that you will say so. If I schedule it and run in CHIP thread it stops working ... It looks like scheduling takes some time, so factory reset comes too fast. In such case I'm afraid that the delay timer would be necessary.